### ISSUE TITLE
import map inconsistency between lsp and runtime

### ISSUE THREAD

**Issue by Reububble** [denoland/deno#24666]
Version: Deno 1.45.2

The lsp seems to be using import maps differently to the runtime. The following example will run fine (`deno run c.ts`) but lsp finds an error in `c.ts` line 10. `deno check c.ts` finds no errors. This used to be fine in version 1.44.4.

<details><summary>folder listing</summary>

```
.
├─deno.jsonc
├─c.ts
├─a
│ ├─deno.jsonc
│ ├─a.ts
│ └─s
│   └─src
│     └─s.ts
└─b
  ├─deno.jsonc
  ├─b.ts
  └─s
    └─src
      └─s.ts
```
</details>

<details><summary>[a/b]/s/src/s.ts </summary>

```
export const s = Symbol("s");

export type S = {
  [s]: number;
};
```
</details>

<details><summary>a/a.ts</summary>

```
import { S, s } from "s/s.ts";

export const a = {
  [s]: 1,
} satisfies S;
```
</details>

<details><summary>b/b.ts</summary>

```
import { S, s } from "s/s.ts";

export const b = {
  [s]: 2,
} satisfies S;
```
</details>

<details><summary>[a/b]/deno.jsonc</summary>

```
{
  "imports": {
    "s/": "./s/src/"
  }
}
```
</details> 

<details><summary>c.ts</summary>

```
import { a } from "a/a.ts";
import { b } from "b/b.ts";
import { S, s } from "s/s.ts";

function xs(x: S) {
  console.log(x[s]);
}

xs(a);
xs(b); // LSP Error in Deno 1.45.0...
/* 
Argument of type '{ [s]: number; }' is not assignable to parameter of type 'S'.
  Property '[s]' is missing in type '{ [s]: number; }' but required in type 'S'.deno-ts(2345)
s.ts(4, 3): '[s]' is declared here.
*/
```
</details>

<details><summary>deno.jsonc</summary>

```
{
  "imports": {
    "a/": "./a/",
    "b/": "./b/",
    "s/": "./a/s/src/"
  }
}
```
</details>

**Comment by nayeemrmn** [comment]
This should be fixed in 1.45.3. Please try it out and comment if it's not

**Comment by Reububble** [comment]
@nayeemrmn The issue persists in 1.45.3

**Comment by dsherret** [comment]
**TLDR**: This isn't a bug. The CLI only discovers the root deno.json file, but the LSP discovers all deno.json files.

The reason this occurs, is because in the lsp it treats the files under each deno.json as being owned by that config. If you were running from the root directory, essentially, the LSP creates an import map that looks like the following:

```
{
  "imports": {
    "a/": "./a/",
    "b/": "./b/",
    "s/": "./a/s/src/"
  },
  "scopes": {
    "./a/": {
      "s/": "./a/s/src/"
    },
    "./b/": {
      "s/": "./b/s/src/"
    }
  }
}
```

Which you can see leads to the same error message that you're encountering in the import map:

```
> deno check c.ts
Check file:///V:/scratch/c.ts
error: TS2345 [ERROR]: Argument of type '{ [s]: number; }' is not assignable to parameter of type 'S'.
  Property '[s]' is missing in type '{ [s]: number; }' but required in type 'S'.
xs(b); // LSP Error in Deno 1.45.0...
   ^
    at file:///V:/scratch/c.ts:10:4

    '[s]' is declared here.
      [s]: number;
      ~~~
        at file:///V:/scratch/a/s/src/s.ts:4:3
```

This is because the LSP discovers all deno.json files, but the deno CLI doesn't. When running `c.ts`, the CLI only discovers the root deno.json file and so it resolves everything with a `s/` specifier to the `"./a/s/src/"` folder.

Another way you can reproduce this behaviour in the CLI, is to use workspaces, which allows you to tell the CLI about the other deno.json files, which will emulate the behaviour in the LSP. For example, adding a workspace field to the root deno.json will also yield the same type checking error on the CLI:

```json
{
  "workspace": {
    "members": [
      "./a",
      "./b"
    ]
  },
  "imports": {
    "a/": "./a/",
    "b/": "./b/",
    "s/": "./a/s/src/"
  }
}
```

**Comment by nayeemrmn** [comment]
To summarise again, you have to add a `workspace` field to unify resolution across a project spanning multiple `deno.json` scopes, otherwise you may come across differences between CLI and LSP.
